common/src/main/java/io/netty/util/NetUtil.java

```diff
@@ -397,48 +381,54 @@ public static boolean isIpV6AddressesPreferred() {
         return null;
     }
 
+    /**
+     * Convert ASCII hexadecimal character to the {@code int} value.
+     * Unlike {@link Character#digit(char, int)}, returns {@code 0} if character is not a HEX-represented.
+     */
     private static int getIntValue(char c) {
-        switch (c) {
-            case '0':
-                return 0;
-            case '1':
-                return 1;
-            case '2':
-                return 2;
-            case '3':
-                return 3;
-            case '4':
-                return 4;
-            case '5':
-                return 5;
-            case '6':
-                return 6;
-            case '7':
-                return 7;
-            case '8':
-                return 8;
-            case '9':
-                return 9;
+        if (c >= '0' && c <= '9') {
+            return c - '0';
         }
-
-        c = Character.toLowerCase(c);
-        switch (c) {
-            case 'a':
-                return 10;
-            case 'b':
-                return 11;
-            case 'c':
-                return 12;
-            case 'd':
-                return 13;
-            case 'e':
-                return 14;
-            case 'f':
-                return 15;
+        if (c >= 'A' && c <= 'F') {
+            // 0xA - a start value in sequence 'A'..'F'
+            return c - 'A' + 0xA;
+        }
+        if (c >= 'a' && c <= 'f') {
+            // 0xA - a start value in sequence 'a'..'f'
+            return c - 'a' + 0xA;
         }
         return 0;
     }
 
+    private static int decimalDigit(String str, int pos) {
+        return str.charAt(pos) - '0';
+    }
+
+    private static byte ipv4WordToByte(String ip, int from, int toExclusive) {
+        int ret = decimalDigit(ip, from);
```

Comment:
Exception in the default case does not affect the result.
Variant with `if(Len == 3) {} else if (Len == 2)` is a bit better for the len == 3.

For the code
```java
        int len = toExclusive - from;
        if (len == 3) {
            return (byte) (decimalDigit(ip, from) * 100 + decimalDigit(ip, from + 1) * 10 + decimalDigit(ip, from + 2));
        }
        if (len == 2) {
            return (byte) (decimalDigit(ip, from) * 10 + decimalDigit(ip, from + 1));
        }
        return (byte) decimalDigit(ip, from);
```
result:
```
IpV4ToBytesBenchmark.ipv4ToBytesNew        127.0.0.1  thrpt    6  71,670 ± 1,757  ops/us
IpV4ToBytesBenchmark.ipv4ToBytesNew  255.255.255.255  thrpt    6  57,369 ± 2,962  ops/us
IpV4ToBytesBenchmark.ipv4ToBytesNew          1.1.1.1  thrpt    6  91,334 ± 0,775  ops/us
```

If do a check `len == 1` firstly
```java
        int len = toExclusive - from;
        if (len == 1) {
            return (byte) decimalDigit(ip, from);
        }
        if (len == 2) {
            return (byte) (decimalDigit(ip, from) * 10 + decimalDigit(ip, from + 1));
        }
        return (byte) (decimalDigit(ip, from) * 100 + decimalDigit(ip, from + 1) * 10 + decimalDigit(ip, from + 2));
```
results:
```java
IpV4ToBytesBenchmark.ipv4ToBytesNew        127.0.0.1  thrpt    6   72,652 ± 1,706  ops/us
IpV4ToBytesBenchmark.ipv4ToBytesNew  255.255.255.255  thrpt    6   48,773 ± 4,266  ops/us
IpV4ToBytesBenchmark.ipv4ToBytesNew          1.1.1.1  thrpt    6  101,944 ± 4,092  ops/us
```